staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/interface.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"time"
+)
+
+// QueueSetFactory knows how to make QueueSet objects.  The request
+// management filter makes a QueueSet for each priority level.  The
+// clock, and any other test-facilitating infrastructure, to use in
+// each QueueSet is known to the factory, the client does not need to
+// pass this stuff in each call to NewQueueSet.
+type QueueSetFactory interface {
+	NewQueueSet(name string, concurrencyLimit, numQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) (QueueSet, error)
+}
+
+// QueueSet is the abstraction for the queuing and dispatching
+// functionality of one non-exempt priority level.  It covers the
+// functionality described in the "Assignment to a Queue", "Queuing",
+// and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type QueueSet interface {
+	// SetConfiguration updates the configuration
+	SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) error
+
+	// Quiesce controls whether this system is quiescing.  Passing a
+	// non-nil handler means the system should become quiescent, a nil
+	// handler means the system should become non-quiescent.  A call
+	// to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.  If all the queues have no requests
+	// waiting nor executing while the system is quiescent then the
+	// handler will eventually be called with no locks held (even if
+	// the system becomes non-quiescent between the triggering state
+	// and the required call).
+	//
+	// The filter uses this for a priority level that has become
+	// undesired, setting a handler that will cause the priority level
+	// to eventually be removed from the filter if the filter still
+	// wants that.  If the filter later changes its mind and wants to
```

Comment:
Is this for the shutdown case? Why is it so complicated? I guess I'd have expected to have a Shutdown() method, which blocks until all queues are empty (returning errors to new requests). I'd expect the adjustment system, whatever that is, to first remove a QueueSet from the active sets, then call Shutdown(), then drop the reference for the GC to clean up.

Reply:
Have a look at what we say in the KEP about deleting a PriorityLevelConfiguration object; that's in the penultimate paragraph of https://github.com/kubernetes/enhancements/blob/483a91a036d385a4dc26b81936c617f49c6ad8cb/keps/sig-api-machinery/20190228-priority-and-fairness.md#reaction-to-configuration-changes :

> When a request priority configuration object is deleted, in a given apiserver the corresponding implementation objects linger until all the queues of that priority level are empty. A FlowSchema associated with one of these lingering undesired priority levels matches no requests.

What's not explicitly written there, but I think implied, is that if a PriorityLevelConfiguration object is deleted and then later, before all its queues have drained, another PriorityLevelConfiguration object is created with the same name, then the lingering queues become associated with the new configuration.  Implementing this means that deletion of a PriorityLevelConfiguration is not simply handled by taking it out of the internal data structure (`requestManagerState`, at https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/reqmgmt.go#L69-L79) and forgetting about it.  Rather, the lingering priority level is retained (yet semi-disabled so that FlowSchemas referencing it direct no more requests there) until its queues all drain or it becomes fully enabled again.  Rather than have a synchronous call to wait for either of those events there is a callback that gets registered to be called if and when the queues are all drained and the priority level is still semi-disabled.

You could say this is a level-based design rather than an edge-based design.  There is a callback registered while the priority level is semi-disabled, rather than a goroutine forked when the priority level becomes semi-disabled.

As for the related question of why Wait returns a `quiescent bool`, that has to do with minimizing the cross-section of locks.  In this design each QueueSet has a mutex and those are the only mutexes.  How does the configuration controller synchronize with the request handling?  Using `sync/atomic` Load and Store.  The configuration controller occasionally Stores an immutable `requestManagerState` (i.e., whenever it finishes digesting the configuration).  The request handler has a loop that starts with a Load of the current `requestManagerState`.  It is possible that, between that Load and the call to Wait on the QueueSet of the relevant priority level, the config controller has digested the deletion of the corresponding PriorityLevelConfiguration object and made the QueueSet quiescent.

Reply:
I would have expected disabled PLs to still execute their remaining queued requests. So I still don't understand why Wait needs to return the quiescent flag.

> The request handler has a loop that starts with a Load of the current requestManagerState. It is possible that, between that Load and the later call to Wait on the QueueSet of the relevant priority level, the config controller has digested some more config ...

What is wrong with just permitting the request to go through that queueset anyway in this condition?

On the name: I think "draining" or "disabled" are much better words for this condition than quiescent. "quiescent" to me means a state has been reached in which no further action is necessary; if the queues are not empty, then the queueset isn't yet quiescent.

Reply:
The functional question is discussed in another conversation, on Wait.

As for the name, I see it was a mistake to use "quiescent"; I really meant "quiescing".  That has the right connotation of moving towards that terminal state but not being there yet.  "Draining" is not totally bad, but has the disadvangate that under normal conditions queues could be said to be draining out one end while being filled at the other end.

Reply:
Perhaps it would make sense to change the text to be `s/Quiesce/Drain` and `s/quiescent/draining` in this case?

Reply:
How about s/quiescent/quiescing/ ?

Reply:
"Quiescent" is confusing, as previously discussed; can we replace it everywhere with "quiescing"?

Reply:
I can live with quiescing as we discussed; the current text hasn't been updated, though?

Reply:
Updated - s/quiescent/quiescing/.  Do  the statements - `become quiescing`, `become non-quiescing` still seem reasonable with this change?

Reply:
How about replacing the first paragraph of the comment with something like the following?

> // Quiesce controls whether the QueueSet is operating normally or is quiescing.
> // A quiescing QueueSet drains as normal but does not admit any
> // new requests. Passing a non-nil handler means the system should
> // be quiescing, a nil handler means the system should operate
> // normally. A call to Wait while the system is quiescing
> // will be rebuffed by returning `tryAnother=true`.  If all the
> // queues have no requests waiting nor executing while the system
> // is quiescing then the handler will eventually be called with no
> // locks held (even if the system becomes non-quiescing between the
> // triggering state and the required call).

Reply:
Updated both entries to the suggested text